File: hw2/q1.cpp
#include <iostream>
#include <vector>

class Stack {
    public:
        std::vector<int> nums; // empty vector
        
        /* Push item onto stack */
        void push(int val) {
            nums.push_back(val);
        }

        /**
         * Remove most recently added item from stack and return it. 
         * Assumes stack is non-empty. 
         */
        int pop() {
            int last_val = nums.back();
            nums.pop_back();
            return last_val;
        }

        /** 
         * Returns most recently added item from stack without removing it. 
         * Assumes stack is non-empty.
         */
        int peek() {
            return nums.back();
        }
};

int main() {
    std::cout << "test 1" << std::endl;
    Stack st;
    st.push(3);
    std::cout << st.peek() << std::endl;
    st.push(4);
    st.push(5);
    std::cout << st.pop() << std::endl;
    std::cout << st.peek() << std::endl;
    st.push(6);
    std::cout << st.pop() << std::endl;

    std::cout << "\ntest 2" << std::endl;
    Stack st2;
    st2.push(1);
    st2.push(2);
    st2.push(3);
    std::cout << st2.pop() << std::endl;
    std::cout << st2.pop() << std::endl;
    std::cout << st2.pop() << std::endl;

    std::cout << "\ntest 3" << std::endl;
    Stack st3;
    st3.push(1);
    st3.push(2);
    std::cout << st3.peek() << std::endl;
    std::cout << st3.peek() << std::endl;

    std::cout << "\ntest 4" << std::endl;
    Stack st4;
    st4.push(1);
    std::cout << st4.peek() << std::endl;
    std::cout << st4.pop() << std::endl;
    st4.push(2);
    std::cout << st4.peek() << std::endl;
    st4.push(5);
    std::cout << st4.peek() << std::endl;
    st4.push(10);
    std::cout << st4.pop() << std::endl;
    std::cout << st4.pop() << std::endl;
    std::cout << st4.pop() << std::endl;

    return 0;
}

File: hw2/q1.py
class Stack:
    def __init__(self):
        self.nums = [] # stack initialized as empty

    def push(self, val):
        ''' Adds an item onto the stack '''
        self.nums.append(val)

    def pop(self):
        ''' Removes most recently added item from stack and returns it. Assumes stack is not empty. '''
        return self.nums.pop()

    def peek(self):
        ''' Returns most recently added item from stack without removing it. Assumes stack is not empty. '''
        return self.nums[-1]

if __name__ == '__main__':
    print("test 1")
    st = Stack()
    st.push(3)
    print(st.peek())
    st.push(4)
    st.push(5)
    print(st.pop())
    print(st.peek())
    st.push(6)
    print(st.pop())

    print("\ntest 2")
    st2 = Stack()
    st2.push(1)
    st2.push(2)
    st2.push(3)
    print(st2.pop())
    print(st2.pop())
    print(st2.pop())

    print("\ntest 3")
    st3 = Stack()
    st3.push(1)
    st3.push(2)
    print(st3.peek())
    print(st3.peek())

    print("\ntest 4")
    st4 = Stack()
    st4.push(1)
    print(st4.peek())
    print(st4.pop())
    st4.push(2)
    print(st4.peek())
    st4.push(5)
    print(st4.peek())
    st4.push(10)
    print(st4.pop())
    print(st4.pop())
    print(st4.pop())